controllers/helmchart_controller.go

```diff
@@ -316,31 +331,111 @@ func (r *HelmChartReconciler) getChartRepositoryWithArtifact(ctx context.Context
-			resetStatus = true
-		}
+func (r *HelmChartReconciler) reconcileFromGitRepository(ctx context.Context,
+	repository sourcev1.GitRepository, chart sourcev1.HelmChart) (sourcev1.HelmChart, error) {
+	// create tmp dir
+	tmpDir, err := ioutil.TempDir("", fmt.Sprintf("%s-%s", chart.Namespace, chart.Name))
+	if err != nil {
+		err = fmt.Errorf("tmp dir error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 	}
+	defer os.RemoveAll(tmpDir)
 
-	// set initial status
-	if len(chart.Status.Conditions) == 0 {
-		resetStatus = true
+	// open file
+	f, err := os.Open(repository.GetArtifact().Path)
+	if err != nil {
+		err = fmt.Errorf("artifact open error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+
+	// extract artifact files
+	if _, err = untar.Untar(f, tmpDir); err != nil {
+		err = fmt.Errorf("artifact untar error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+
+	// ensure configured path is a chart directory
+	chartPath := path.Join(tmpDir, chart.Spec.Chart)
+	if _, err := chartutil.IsChartDir(chartPath); err != nil {
+		err = fmt.Errorf("chart path error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+
+	// read chart metadata
+	chartMetadata, err := chartutil.LoadChartfile(path.Join(chartPath, chartutil.ChartfileName))
+	if err != nil {
+		err = fmt.Errorf("load chart metadata error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 	}
 
-	return resetStatus, sourcev1.HelmChartStatus{
-		Conditions: []sourcev1.SourceCondition{
-			{
-				Type:               sourcev1.ReadyCondition,
-				Status:             corev1.ConditionUnknown,
-				Reason:             sourcev1.InitializingReason,
-				LastTransitionTime: metav1.Now(),
-			},
-		},
+	// return early on unchanged chart version
+	if chart.Status.Artifact != nil && chartMetadata.Version == chart.Status.Artifact.Revision {
+		return chart, nil
 	}
+
+	artifact := r.Storage.ArtifactFor(chart.Kind, chart.ObjectMeta.GetObjectMeta(),
+		fmt.Sprintf("%s-%s.tgz", chartMetadata.Name, chartMetadata.Version), chartMetadata.Version)
+
+	// create artifact dir
+	err = r.Storage.MkdirAll(artifact)
+	if err != nil {
+		err = fmt.Errorf("unable to create artifact directory: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+
+	// acquire lock
+	unlock, err := r.Storage.Lock(artifact)
+	if err != nil {
+		err = fmt.Errorf("unable to acquire lock: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+	defer unlock()
+
+	// package chart
+	pkg := action.NewPackage()
+	pkg.Destination = artifact.Path
+	_, err = pkg.Run(chartPath, nil)
```

Comment:
How is this dealing with local dependencies and password protected Helm repos?

Reply:
It is not, it is expected (at least for now) that they are checked in to Git and made available in the `charts/` directory. Otherwise we will run into issues like https://github.com/fluxcd/helm-operator/blob/0ed00cd43a967207fca1f18c10121bd73accb3d4/pkg/helm/v3/dependency.go#L34, or we would have to write our own download manager.